File: src/parser.js
const QUALITY_RULES = [
  [/\b(2160p|4k|uhd)\b/i, '4K'],
  [/\b1080p\b/i, '1080p'],
  [/\b720p\b/i, '720p'],
  [/\b480p\b/i, '480p'],
  [/\bcam\b/i, 'CAM'],
]

const SOURCE_RULES = [
  [/\bremux\b/i, 'Remux'],
  [/\b(bluray|bdrip|brrip)\b/i, 'BluRay'],
  [/\bweb[-.\s]?dl\b/i, 'WEB-DL'],
  [/\bwebrip\b/i, 'WEBRip'],
  [/\bhdtv\b/i, 'HDTV'],
  [/\bcam\b/i, 'CAM'],
]

const HDR_RULES = [
  [/\b(dv|dolby[.\s]?vision)\b/i, 'DV'],
  [/\bhdr10\+?\b/i, 'HDR10'],
  [/\bhdr\b/i, 'HDR'],
]

const CODEC_RULES = [
  [/\b(x265|hevc|h[.\s]?265)\b/i, 'x265'],
  [/\b(x264|h[.\s]?264)\b/i, 'x264'],
  [/\bav1\b/i, 'AV1'],
]

const AUDIO_RULES = [
  [/\bmulti\b/i, 'MULTI'],
  [/\b(truefrench|french|vff|vfq)\b/i, 'FRENCH'],
  [/\bvostfr\b/i, 'VOSTFR'],
  [/\b(english|eng)\b/i, 'ENG'],
]

function firstMatch(rules, title) {
  for (const [re, label] of rules) {
    if (re.test(title)) return label
  }
  return null
}

export function parseTorrentTitle(title) {
  const t = title || ''
  return {
    quality: firstMatch(QUALITY_RULES, t),
    source: firstMatch(SOURCE_RULES, t),
    hdr: firstMatch(HDR_RULES, t),
    codec: firstMatch(CODEC_RULES, t),
    audio: firstMatch(AUDIO_RULES, t),
  }
}

File: internal/parser/parser.go
package parser

import "regexp"

type Parsed struct {
	Quality string
	Source  string
	HDR     string
	Codec   string
	Audio   string
}

type rule struct {
	re    *regexp.Regexp
	label string
}

var qualityRules = []rule{
	{regexp.MustCompile(`(?i)\b(2160p|4k|uhd)\b`), "4K"},
	{regexp.MustCompile(`(?i)\b1080p\b`), "1080p"},
	{regexp.MustCompile(`(?i)\b720p\b`), "720p"},
	{regexp.MustCompile(`(?i)\b480p\b`), "480p"},
	{regexp.MustCompile(`(?i)\bcam\b`), "CAM"},
}

var sourceRules = []rule{
	{regexp.MustCompile(`(?i)\bremux\b`), "Remux"},
	{regexp.MustCompile(`(?i)\b(bluray|bdrip|brrip)\b`), "BluRay"},
	{regexp.MustCompile(`(?i)\bweb[-.\s]?dl\b`), "WEB-DL"},
	{regexp.MustCompile(`(?i)\bwebrip\b`), "WEBRip"},
	{regexp.MustCompile(`(?i)\bhdtv\b`), "HDTV"},
	{regexp.MustCompile(`(?i)\bcam\b`), "CAM"},
}

var hdrRules = []rule{
	{regexp.MustCompile(`(?i)\b(dv|dolby[.\s]?vision)\b`), "DV"},
	{regexp.MustCompile(`(?i)\bhdr10\+?\b`), "HDR10"},
	{regexp.MustCompile(`(?i)\bhdr\b`), "HDR"},
}

var codecRules = []rule{
	{regexp.MustCompile(`(?i)\b(x265|hevc|h[.\s]?265)\b`), "x265"},
	{regexp.MustCompile(`(?i)\b(x264|h[.\s]?264)\b`), "x264"},
	{regexp.MustCompile(`(?i)\bav1\b`), "AV1"},
}

var audioRules = []rule{
	{regexp.MustCompile(`(?i)\bmulti\b`), "MULTI"},
	{regexp.MustCompile(`(?i)\b(truefrench|french|vff|vfq)\b`), "FRENCH"},
	{regexp.MustCompile(`(?i)\bvostfr\b`), "VOSTFR"},
	{regexp.MustCompile(`(?i)\b(english|eng)\b`), "ENG"},
}

func firstMatch(rules []rule, title string) string {
	for _, r := range rules {
		if r.re.MatchString(title) {
			return r.label
		}
	}
	return ""
}

func Parse(title string) Parsed {
	return Parsed{
		Quality: firstMatch(qualityRules, title),
		Source:  firstMatch(sourceRules, title),
		HDR:     firstMatch(hdrRules, title),
		Codec:   firstMatch(codecRules, title),
		Audio:   firstMatch(audioRules, title),
	}
}
